src/lib/support/PersistentArray.h

```diff
@@ -0,0 +1,316 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/support/PersistentData.h>
+
+namespace chip {
+
+enum class PersistentTags : uint8_t
+{
+    kCount   = 1,
+    kList    = 2,
+    kEntries = 3,
+};
+
+template <size_t kMaxSerializedSize, size_t kMaxListSize, typename EntryType>
+struct PersistentArray : public PersistentData<kMaxSerializedSize>
+{
+
+    PersistentArray(PersistentStorageDelegate * storage) : PersistentData<kMaxSerializedSize>(storage), mLimit(kMaxListSize) {}
+
+    size_t Limit() { return this->mLimit; }
+    uint16_t Count() { return this->mCount; }
+    virtual void Clear(EntryType & entry)                                                = 0;
+    virtual CHIP_ERROR Serialize(TLV::TLVWriter & writer, const EntryType & entry) const = 0;
+    virtual CHIP_ERROR Deserialize(TLV::TLVReader & reader, EntryType & entry)           = 0;
+
+    virtual CHIP_ERROR Copy(EntryType & target, const EntryType & source)
+    {
+        target = source;
+        return CHIP_NO_ERROR;
+    }
+
+    virtual bool Compare(const EntryType & a, const EntryType & b) const { return a == b; }
+
+    void Clear() override
+    {
+        // Clear entries
+        for (size_t i = 0; i < this->mLimit; ++i)
+        {
+            EntryType & e = mEntries[i];
+            this->Clear(e);
+        }
+        this->mCount = 0;
+    }
+
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override
+    {
+        TLV::TLVType container, list;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, container));
+
+        // Count
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(PersistentTags::kCount), static_cast<uint16_t>(this->mCount)));
+
+        // Entries
+        ReturnErrorOnFailure(writer.StartContainer(TLV::ContextTag(PersistentTags::kList), TLV::kTLVType_List, list));
+        for (size_t i = 0; (i < this->mCount) && (i < this->mLimit); ++i)
+        {
+            ReturnErrorOnFailure(Serialize(writer, mEntries[i]));
+        }
+        ReturnErrorOnFailure(writer.EndContainer(list));
+
+        return writer.EndContainer(container);
+    }
+
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override
+    {
+        ReturnErrorOnFailure(reader.Next(TLV::AnonymousTag()));
+        VerifyOrReturnError(TLV::kTLVType_Structure == reader.GetType(), CHIP_ERROR_INTERNAL);
```

Comment:
Why not pass your desired type to Next(), then, and have the TLV machinery do that check for you instead of doing it manually?